p2p/host/basic/basic_host.go

```diff
@@ -631,10 +631,24 @@ func (h *BasicHost) NewStream(ctx context.Context, p peer.ID, pids ...protocol.I
 		}, nil
 	}
 
-	selected, err := msmux.SelectOneOf(pidStrings, s)
-	if err != nil {
+	// Negotiate the protocol in the background, obeying the context.
+	var selected string
+	errCh := make(chan error, 1)
+	go func() {
+		selected, err = msmux.SelectOneOf(pidStrings, s)
+		errCh <- err
+	}()
```

Comment:
Don't like the extra goroutine, but it's gotta be done.

Reply:
Ideally, we'd pass the context down. But we'd need to goroutine at _some_ level, eventually.